sound/soc/sof/sof-client.c

```diff
@@ -107,3 +107,13 @@ struct dentry *sof_client_get_debugfs_root(struct sof_client_dev *cdev)
 	return cdev->sdev->debugfs_root;
 }
 EXPORT_SYMBOL_NS_GPL(sof_client_get_debugfs_root, SND_SOC_SOF_CLIENT);
+
+/*
+ * DMA buffer alloc fails when using the client device. Use the SOF core device instead.
+ * This will be needed for clients other than the probes client device as well.
```

Comment:
I can't recall why this was required. @ranj063 did we ever root cause this?

Or is this saying that clients can only deal with IPC, and all DMA allocation is handled by the core?

Reply:
If the memory is shared between the CPU and a device then the memory should be allocated via the DMA API to make sure that the coherency (or the lack of it) is handled correctly (flush/invalidate).
Only the parent device is a real device, so we need to use that for allocations.

Reply:
I am having difficulties connecting the dots with the SPIB work and more generically not sure how streaming would be handled by a client.

if the client deals with writes (playback) and reads (capture), how would we program this information back into the hardware registers that are owned by the core, or conversely how would the core inform the client of an xrun.

Reply:
This SOF client auxbus representation can only be applied to pure IPC clients.
I still get confused time to time when I re-read the auxiliary bus documentation, but I usually arrive to the pow that unlike with MFD drivers for auxbus devices should not touch or have access to parent (in our case core's) regmap or register space.
Which is a bit odd as there are devices out there where one register map contains registers for different functionalities, like PMIC with integrated audio codec.

We should have devices on the auxbus out from the core where the sole functionality is via the firmware, it starts to be a bit on the blurry side when you have MMR since you can say that then we have a real device with register map.

Keeping that in mind we also need to see the implications of having auxbus for some functionality and how that one would interact with other subsystems (HDMI?).

In my view we should not force the use auxbus just for the sake of using auxbus.

Reply:
I will spend more days to dig into this, I have a feeling that I'm missing some vital information.

Reply:
right, and that's were I am confused. 

part of the probe functionality is to send a bunch of IPC. that's fine and aligned with what we wanted the auxbus to do.

the data handling part is less clear to me. There's a snd_soc_dai_ops in a dai driver used by the core. I can't figure out who allocates the DMA buffer.

Reply:
Hrm, the dai driver should not allocate the DMA memory, at least not in ASoC.
It is still a bit unclear for me what the dai driver is in SOF. Isn't it so that the firmware is in control of the real dai and bus and signals?

In general the PCM driver or lib should allocate the DMA memory for data. In classical ASoC most of are using the dmaengine-pcm as a generic layer and this generic layer here in SOF is the IPC.
The IPC knows how to move data to/from the firmware, it might be a shared buffer with mailbox signaling, it can be a stream port, it can be an SPI bus or SoundWire or even ethernet (in theory).
The data allocation should use the device for this interface, but the data allocation itself is somewhere else.

I think this function has been added here because the probe provides custom interface to send and receive arbitrary data as well and it needs to allocate the memory using the correct device?

Reply:
There's probably a bit of API abuse here. 
- The DAIs that are exposed by the drivers are related to HDaudio, SSP and DMIC. 
- What the probes do with this batch isn't clear to me at all. I think the dai is meant as 'dai associated to the compressed stream', just like in topology we have DAI widgets for each PCM stream. 

I wonder if this was added this way so that we didn't have to add it in topology files?
@ranj063 any memories of this work from last year?